File: backend/src/main/java/com/kiwi/features/nodes/data/NodesDataMapper.java
package com.kiwi.features.nodes.data;

import java.util.List;

public class NodesDataMapper {

    public static NodesDomain toDomain(NodesPersistence node, UserNodeStatusPersistence userStatus) {
        return NodesDomainFactory.create(node, userStatus);

    }

    public static NodesDTO toDTO(NodesDomain node) {
        return new NodesDTO(
                node.getId(),
                node.getStatus().name(),
                node.getIcon(),
                node.getPrice(),
                node.getCordX(),
                node.getCordY(),
                node.getEventOnExecution(),
                node.getName(),
                node.getDisplayName(),
                node.getConnectedNodeIds(),
                node.getMapId(),
                node.isFirstNodeOfMap()
        );
    }

    public static NodesDTO toDTO(NodesPersistence node, UserNodeStatusPersistence status) {
        List<Long> connectedNodeIds = node.getOutgoingEdges()
                .stream()
                .map(edge -> edge.getToNode().getId())
                .toList();

        return new NodesDTO(
                node.getId(),
                status.getStatus().name(),
                node.getIcon(),
                node.getPrice(),
                node.getCordX(),
                node.getCordY(),
                node.getEventOnExecution(),
                node.getName(),
                node.getDisplayName(),
                connectedNodeIds,
                node.getMapId(),
                node.isFirstNodeOfMap()
        );
    }

    public static UserNodeStatusPersistence toPersistence(Long userId, NodesDomain node) {
        if (node.getStatus() == NodeStatus.INACCESSIBLE) {
            return null;
        }

        UserNodeStatusKey key = new UserNodeStatusKey(userId, node.getId());

        UserNodeStatusPersistence entity = new UserNodeStatusPersistence();
        entity.setId(key);
        entity.setStatus(node.getStatus());
        return entity;
    }
}

File: mobile/app/src/main/java/com/bellako/kiwi/features/nodes/data/NodesDataMapper.kt
package com.bellako.kiwi.features.nodes.data

object NodesDataMapper {
    fun toDomain(dto: NodesDTO): NodesDomain =
        NodesDomain(
            id = dto.id,
            icon = dto.icon,
            status = NodeStatus.valueOf(dto.status),
            price = dto.price,
            cordX = dto.cordX,
            cordY = dto.cordY,
            name = dto.name,
            displayName = dto.displayName,
            connectedNodeIds = dto.connectedNodeIds,
            mapId = dto.mapId,
            onExecutionEvent = dto.onExecutionEvent,
            onExecutionEntityId = dto.onExecutionEntityId,
            transitionStyle = NodeTransitionStyle.valueOf(dto.transitionStyle),
        )

    fun toDTO(domain: NodesDomain): NodesDTO =
        NodesDTO(
            id = domain.id,
            icon = domain.icon,
            status = domain.status.name,
            price = domain.price,
            cordX = domain.cordX,
            cordY = domain.cordY,
            name = domain.name,
            displayName = domain.displayName,
            connectedNodeIds = domain.connectedNodeIds,
            mapId = domain.mapId,
            onExecutionEvent = domain.onExecutionEvent,
            onExecutionEntityId = domain.onExecutionEntityId,
            transitionStyle = domain.transitionStyle.name,
        )
}
